src/rust/engine/src/tasks.rs

```diff
@@ -59,6 +63,39 @@ impl Tasks {
     }
   }
 
+  pub fn all_product_types(&self) -> Vec<TypeConstraint> {
+    self.all_rules().iter().map(|t| t.product).collect()
+  }
+
+  pub fn is_singleton_task(&self, sought_task: &Task) -> bool {
+    self.singletons.values().any(|tasks| tasks.iter().any(|t| t == sought_task))
+  }
+
+  pub fn is_intrinsic_task(&self, sought_task: &Task) -> bool {
+    self.intrinsics.values().any(|tasks| tasks.iter().any(|t| t == sought_task))
+  }
+
+  pub fn all_rules(&self) -> Vec<Task> {
+    let mut result: Vec<Task> = Vec::new();
```

Comment:
Another way to accomplish method would be:

let rules: Vec<_> = self.tasks.values().chain(self.singletons.values()).chain(self.singletons.values()).collect();